qa/tasks/cephfs/test_quiesce.py

```diff
@@ -0,0 +1,627 @@
+S_IFDIR  =  0o040000
+S_IFCHR  =  0o020000
+S_IFIFO  =  0o010000
+S_ISUID  =  0o004000
+S_ISGID  =  0o002000
+S_ISVTX  =  0o001000
+
+class QuiesceTestCase(CephFSTestCase):
+    """
+    Test case for quiescing subvolumes.
+    """
+
+    CLIENTS_REQUIRED = 2
+    MDSS_REQUIRED = 1
+
+    QUIESCE_SUBVOLUME = "subvol_quiesce"
+
+    def setUp(self):
+        super().setUp()
+        self.run_ceph_cmd(f'fs subvolume create {self.fs.name} {self.QUIESCE_SUBVOLUME} --mode=777')
+        p = self.run_ceph_cmd(f'fs subvolume getpath {self.fs.name} {self.QUIESCE_SUBVOLUME}', stdout=StringIO())
+        self.mntpnt = p.stdout.getvalue().strip()
+        self.subvolume = self.mntpnt
+
+    def tearDown(self):
+        # restart fs so quiesce commands clean up and commands are left unkillable
+        self.fs.fail()
+        self.fs.set_joinable()
+        self.fs.wait_for_daemons()
+        super().tearDown()
+
+    def _configure_subvolume(self):
+        for m in self.mounts:
+            m.umount_wait()
+        for m in self.mounts:
+            m.update_attrs(cephfs_mntpt = self.mntpnt)
+            m.mount()
+
+    CLIENT_WORKLOAD = """
+        set -ex
+        pushd `mktemp -d -p .`
+        cp -a /usr .
+        popd
+    """
+    def _client_background_workload(self):
+       for m in self.mounts:
+           p = m.run_shell_payload(self.CLIENT_WORKLOAD, wait=False, stderr=StringIO())
+           m.background_procs.append(p)
+
+    def _wait_for_quiesce_complete(self, reqid, rank=0, path=None):
+        op = None
+        try:
+            with safe_while(sleep=1, tries=120, action='wait for quiesce completion') as proceed:
+                while proceed():
+                    op = self.fs.get_op(reqid, rank=rank)
+                    log.debug(f"op:\n{op}")
+                    self.assertEqual(op['type_data']['op_name'], 'quiesce_path')
+                    if op['type_data']['flag_point'] in (self.FP_QUIESCE_COMPLETE, self.FP_QUIESCE_COMPLETE_NON_AUTH_TREE):
+                        return op
+        except:
+            log.info(f"op:\n{op}")
+            if path is not None:
+                cache = self.fs.read_cache(path, rank=rank)
+                (fd, path) = tempfile.mkstemp()
+                with os.fdopen(fd, "wt") as f:
+                    f.write(f"{json.dumps(cache, indent=2)}")
+                    log.error(f"cache written to {path}")
+            (fd, path) = tempfile.mkstemp()
+            with os.fdopen(fd, "wt") as f:
+                ops = self.fs.get_ops(locks=True, rank=rank)
+                f.write(f"{json.dumps(ops, indent=2)}")
+                log.error(f"ops written to {path}")
+            raise
+
+    FP_QUIESCE_COMPLETE = 'quiesce complete'
+    FP_QUIESCE_BLOCKED = 'quiesce blocked'
+    FP_QUIESCE_COMPLETE_NON_AUTH = 'quiesce complete for non-auth inode'
+    FP_QUIESCE_COMPLETE_NON_AUTH_TREE = 'quiesce complete for non-auth tree'
+    def _verify_quiesce(self, rank=0, root=None, splitauth=False):
```

Comment:
I'm concerned with this test peeking into the implementation details. Any change to the internal quiesce logic will cause this to fail with false positive given that the IO pause was still achieved

Reply:
Would you agree that what we're really trying to verify here is that it is not possible to acquire any of the important locks while a node is quiesced. However, I don't know how to properly do that from the Python side. In any case, it might be just not the setup to perform this kind of validation, since it's not a unittest on the mds side, but a behavioral test from the FS mount perspective.
Hence my opinion is that we should restrain to only testing the locks indirectly by trying to execute operations that would need to take those locks, like interacting with the mounted fs, or, potentially, sending some direct libcephfs client calls

Reply:
We're testing a few things during verification:

- That every inode in a tree being quiesced has a `quiesce_inode` sub-op. And that no `sub-op` exists for an inode not in the tree.
- That, yes, the locks we expect to be acquired are in-fact acquired by the `quiesce_inode` op. i.e. we did not somehow "succeed" quiesce when there are outstanding ops that have not yet acquired all the necessary ops.
- That final flag_points are as we expect.

It is obviously tied to the implementation but this is not unusual for these types of tests. Unless the locks acquired were to change or new terminal states are added (to flag_points), I don't expect this procedure to change.

Reply:
ack